sklearn/tests/test_metaestimators_metadata_routing.py

```diff
@@ -17,109 +22,295 @@
+    warns_on = metaestimator["warns_on"]
+
+    for method_name in routing_methods:
+        if method_name not in warns_on:
+            # this method is not expected to warn
+            continue
+
+        for key in warns_on[method_name]:
+            val = {"sample_weight": sample_weight, "metadata": metadata}[key]
+            kwargs = {key: val}
+            warn_msg = (
+                "You are passing metadata for which the request values are not"
+                f" explicitly set: {key}. From version 1.4 this results in the"
+                f" following error: [{key}] are passed but are not explicitly set as"
+                f" requested or not for {estimator.__class__.__name__}.{method_name}"
+            )
+            with pytest.warns(FutureWarning, match=re.escape(warn_msg)):
+                method = getattr(instance, method_name)
+                method(X, y, **kwargs)
+
+            if metaestimator.get("preserves_metadata", True):
+                # sanity check that registry is not empty, or else the test
+                # passes trivially
+                assert registry
+                for estimator in registry:
+                    check_recorded_metadata(estimator, method_name, **kwargs)
+            # clear the registry since the check could be different for the next
+            # method being tested
+            registry.clear()
 
 
 @pytest.mark.parametrize(
-    "MultiOutput, Estimator",
-    [
-        (MultiOutputClassifier, ConsumingClassifier),
-        (MultiOutputRegressor, ConsumingRegressor),
-    ],
-    ids=["Classifier", "Regressor"],
+    "metaestimator",
+    METAESTIMATORS,
+    ids=METAESTIMATOR_IDS,
 )
-def test_multioutput_metadata_routing(MultiOutput, Estimator):
-    # Check routing of metadata
-    metaest = MultiOutput(Estimator())
-    warn_msg = (
-        "You are passing metadata for which the request values are not explicitly"
-        " set: sample_weight, metadata."
-    )
-    with pytest.warns(FutureWarning, match=(warn_msg)):
-        metaest.fit(X, y_multi, sample_weight=sample_weight, metadata=metadata)
-        check_recorded_metadata(
-            metaest.estimators_[0],
-            "fit",
-            sample_weight=sample_weight,
-            metadata=metadata,
+def test_error_for_other_methods(metaestimator):
+    # This test complements test_warning_for_indicated_methods but checks for
+    # UnsetMetadataPassedError instead of FutureWarning
+    cls = metaestimator["metaestimator"]
+    estimator = metaestimator["estimator"]()
+    estimator_name = metaestimator["estimator_name"]
+    instance = cls(**{estimator_name: estimator})
+    X = metaestimator["X"]
+    y = metaestimator["y"]
+    routing_methods = metaestimator["routing_methods"]
+    warns_on = metaestimator["warns_on"]
+
+    for method_name in routing_methods:
+        if method_name in warns_on:
+            # this method is expected to warn, not raise
+            continue
```

Comment:
same as the above test, if the method's name is in `warns_on`, it doesn't mean it always warns, it means it warns only for those attributes which are listed there. So we need to test for the other attributes.

Reply:
You're right, the error test now works analogously to the warning test.